api/controllers/UserController.js

```diff
@@ -1323,11 +1309,115 @@ module.exports = {
    *             required: true
    *             description: >-
    *               Should correspond to the endpoint you are interacting with.
+   * responses:
+   *   '204':
+   *     description: >-
+   *       The request to send a password reset email was received and processed
+   *       successfully. NOTE: due to security requirements regarding disclosure
+   *       of email addresses in our system, the response does NOT guarantee the
+   *       email exists, and thus does NOT guarantee that an email was sent.
+   *   '400':
+   *     description: Bad request. See response body for details.
+   * security: []
+   */
+  async resetPasswordEmail(request, reply) {
+    // Look for required params and fail if either are missing.
+    if (!request.payload || !request.payload.email || !request.payload.reset_url) {
+      throw Boom.badRequest('Missing email and/or reset_url parameters in request body.');
+    }
+
+    // Validate app URL.
+    const resetUrl = request.payload.reset_url || '';
+    if (!HelperService.isAuthorizedUrl(resetUrl)) {
+      logger.warn(
+        `[UserController->resetPasswordEmail] reset_url ${resetUrl} is not in allowedDomains list`,
+        {
+          request,
+          security: true,
+          fail: true,
+        },
+      );
+      throw Boom.badRequest('reset_url is invalid');
+    }
+
+    // Lookup user based on the email address. We scan the `emails` array so
+    // that secondary addresses can also receive password resets.
+    const record = await User.findOne({ 'emails.email': request.payload.email.toLowerCase() });
+
+    // No user found.
+    if (!record) {
+      logger.warn(
+        `[UserController->resetPasswordEmail] No user found with email: ${request.payload.email}`,
+        {
+          request,
+          fail: true,
+        },
+      );
+
+      // Send HTTP 204 (empty success response)
+      //
+      // We send this because disclosing the existence (or lack thereof) of an
+      // email address is considered a security hole, per OICT policy.
+      return reply.response().code(204);
+    }
+
+    // Determine whether email is primary. We allow unvalidated primary emails
+    // to accept password resets.
+    const emailIsPrimary = record.email === request.payload.email.toLowerCase();
+
+    // Verify that the email has already been validated
+    // eslint-disable-next-line max-len
+    const secondaryEmailIsValidated = record.emails.some(e => e.email === request.payload.email.toLowerCase() && e.validated === true);
+
+    // IF the email is primary OR it's a __validated__ secondary email
+    // THEN send password reset.
+    if (emailIsPrimary || secondaryEmailIsValidated) {
+      await EmailService.sendResetPassword(record, resetUrl, request.payload.email);
+    } else {
+      logger.warn(
+        `[UserController->resetPasswordEmail] Could not reset password because the secondary email ${request.payload.email} has not been validated.`,
```

Comment:
I agree that it can be a confusing log entry. Luckily, my very next ticket this week is HID-2219 to allow unconfirmed secondary emails to receive password resets (and validate them when the link is used). So the code will get revisited shortly!